tachyon/crypto/sumcheck/multilinear/data_structures.h

```diff
@@ -0,0 +1,167 @@
+    return max_multiplicands == other.max_multiplicands &&
+           num_variables == other.num_variables;
+  }
+  bool operator!=(const PolynomialInfo& other) const {
+    return !operator==(other);
+  }
+  // TODO(ashjeong): For better unit testing
+  // PolynomialInfo Random() {}
+};
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials;
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials<math::MultilinearExtension<Evaluations>> {
+ public:
+  using F = typename math::MultilinearExtension<Evaluations>::Field;
+
+  struct Term {
+    F coefficient;
+    std::vector<size_t> indexes;
+  };
+
+  ListOfProductsOfPolynomials() = default;
+  explicit ListOfProductsOfPolynomials(size_t num_variables)
+      : num_variables_(num_variables) {}
+  ListOfProductsOfPolynomials(ListOfProductsOfPolynomials& other) = default;
+  ListOfProductsOfPolynomials& operator=(ListOfProductsOfPolynomials& other) =
+      default;
+
+  size_t max_evaluations() { return max_evaluations_; }
+  size_t num_variables() { return num_variables_; }
+  std::vector<Term> terms() { return terms_; }
+  std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>
+  flattened_ml_evaluations() {
+    return flattened_ml_evaluations_;
+  }
+
+  // Adds a Term to the list
+  void AddTerm(const F& coefficient,
+               const std::vector<
+                   std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+                   evaluations) {
+    CHECK(!evaluations.empty());
+    size_t num_evaluations = evaluations.size();
+    max_evaluations_ = std::max(max_evaluations_, num_evaluations);
+    std::vector<size_t> indexed_evaluations;
+    indexed_evaluations.reserve(num_evaluations);
+    for (std::shared_ptr<math::MultilinearExtension<Evaluations>> evaluation :
+         evaluations) {
+      CHECK_EQ(evaluation->Degree(), num_variables_);
+      const math::MultilinearExtension<Evaluations>* evaluation_ptr =
+          evaluation.get();
+      if (base::Contains(raw_pointers_lookup_table_, evaluation_ptr)) {
+        indexed_evaluations.push_back(
+            raw_pointers_lookup_table_[evaluation_ptr]);
+      } else {
+        size_t curr_index = flattened_ml_evaluations_.size();
+        flattened_ml_evaluations_.push_back(evaluation);
+        raw_pointers_lookup_table_[evaluation_ptr] = curr_index;
+        indexed_evaluations.push_back(curr_index);
+      }
+    }
+    terms_.push_back({coefficient, indexed_evaluations});
+  }
+
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // ∑ᵢ₌₀ⁿ(coefficientᵢ⋅∏ⱼ₌₀ᵐevaluationᵢⱼ)
+  // where n = total terms, m = number of evaluations per term
+  F Evaluate(std::vector<F>& point) const {
+    CHECK_EQ(point.size(), num_variables_);
+    return std::accumulate(
+        terms_.begin(), terms_.end(), F::Zero(),
+        [this, &point](F& sum, const Term& term) {
+          return sum +=
+                 term.coefficient *
```

Comment:
Certainly, it depends on the expected number of terms. If there are many, please parallelize it.